### ISSUE TITLE
Possible bug in IQN - randomly sampling taus during evaluation?

### ISSUE THREAD

**Issue by emrul** [aai-institute/tianshou#969]
- [X] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [X] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [ ] new feature request
- [X] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [X] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [-] I have mentioned version numbers, operating system and environment, where applicable:
  
Hi, I've observed that during evaluation an IQN policy will output different Q-Values for the same input data. Most of the time the resulting Max Q-Value is the same but sometimes it can differ.

This creates a reproducibility issue for me and I did not think a stock IQN would be stochastic during evaluation.

I think the problem is cause by [this line](https://github.com/thu-ml/tianshou/blob/bf7841078dc5d4357ba8046efb656674af20f54b/tianshou/utils/net/discrete.py#L209):
`taus = torch.rand(batch_size, sample_size, dtype=logits.dtype, device=logits.device)`

It randomly samples taus (which is correct). The trouble is there's no distinction made when `self.training` is `False`. Is this intended and if so, what's the best way to ensure I can get reliable and reproducible results during evaluation?

**Note**: The issue won't arise if you run through an environment with a policy in a given order because the starting seed will be the same on each run. However, if you try to repeat an episode via some other code (as I am doing) you should find the results can differ in a small number of cases.
